dot/core/process.go

```diff
@@ -0,0 +1,292 @@
+}
+
+// BlockResponseMessage 2
+
+// ProcessBlockResponseMessage attempts to validate and add the block to the
+// chain by calling `core_execute_block`. Valid blocks are stored in the block
+// database to become part of the canonical chain.
+func (s *Service) ProcessBlockResponseMessage(msg network.Message) error {
+	log.Debug("[core] received BlockResponseMessage")
+
+	// cast Message to BlockResponseMessage in order to create header
+	blockResponseMessage, ok := msg.(*network.BlockResponseMessage)
+	if !ok {
+		return errors.New("could not cast Message to BlockResponseMessage")
+	}
+
+	// loop through block data in block response message
+	for _, bd := range blockResponseMessage.BlockData {
+		if bd.Header.Exists() {
+
+			header, err := types.NewHeader(
+				bd.Header.Value().ParentHash,
+				bd.Header.Value().Number,
+				bd.Header.Value().StateRoot,
+				bd.Header.Value().ExtrinsicsRoot,
+				bd.Header.Value().Digest,
+			)
+			if err != nil {
+				return err
+			}
+
+			// check authority
+			if s.isAuthority {
+
+				// check if block is from current epoch
+				err = s.verifyCurrentEpoch(header)
+				if err != nil {
+					log.Trace("[core] block is not from current epoch", "error", err)
+				}
+
+				// check if first block and set epoch data for next epoch
+				err = s.handleConsensusDigest(header)
+				if err != nil {
+					log.Trace("[core] block is not first block in epoch", "error", err)
+				}
+			}
+		}
+	}
+
+	// send block response message to syncer
+	s.respOut <- blockResponseMessage
+
+	// TODO: should we wait for syncer to handle message before we check for runtime changes...?
+
+	// check if we need to update the runtime
+	err := s.checkForRuntimeChanges()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// BlockAnnounceMessage 3
+
+// ProcessBlockAnnounceMessage creates a block request message from the block
+// announce messages (block announce messages include the header but the full
+// block is required to execute `core_execute_block`).
+func (s *Service) ProcessBlockAnnounceMessage(msg network.Message) error {
+	log.Debug("[core] received BlockAnnounceMessage")
+
+	blockAnnounceMessage, ok := msg.(*network.BlockAnnounceMessage)
+	if !ok {
+		return errors.New("failed to cast Message to BlockAnnounceMessage")
+	}
+
+	// create header from message
+	header, err := types.NewHeader(
+		blockAnnounceMessage.ParentHash,
+		blockAnnounceMessage.Number,
+		blockAnnounceMessage.StateRoot,
+		blockAnnounceMessage.ExtrinsicsRoot,
+		blockAnnounceMessage.Digest,
+	)
+	if err != nil {
+		return err
+	}
+
+	// check authority
+	if s.isAuthority {
+
+		// check if block is from current epoch
+		err = s.verifyCurrentEpoch(header)
+		if err != nil {
+			log.Trace("[core] block is not from current epoch", "error", err)
+		}
+
+		// check if first block and set epoch data for next epoch
+		err = s.handleConsensusDigest(header)
+		if err != nil {
+			log.Trace("[core] block is not first block in epoch", "error", err)
+		}
+	}
```

Comment:
since this code is the same as above, maybe move it to another function that can be used by ProcessBlockAnnounceMessage and ProcessBlockResponseMessage

Reply:
agreed, will fix